lib/net/winrm/connection.rb

```diff
@@ -0,0 +1,289 @@
+            def read_response(wsmv_message, wait_for_done_state = false)
+              resp_doc = nil
+              until command_done?(resp_doc, wait_for_done_state)
+                logger.debug('[WinRM] Waiting for output...')
+                resp_doc = send_get_output_message(wsmv_message.build)
+                logger.debug('[WinRM] Processing output')
+                read_streams(resp_doc) do |stream|
+                  yield stream, resp_doc
+                end
+              end
+    
+              if command_done?(resp_doc, true)
+                raise EOFError.new('Program terminated')
+              end
+            end
+          end
+    
+          # We create our own empty finalizers because the built-in one triggers a
+          # request using the Rex HTTP client, which segfaults; possibly because it
+          # creates a thread, or something else that is not allowed in a finalizer.
+          # In this situation (observed only when the user quits MSF with active sessions),
+          # we'll just let the shell continue.
+          def remove_finalizer
+          end
+          def add_finalizer
+          end
+    
+          def create_proc
+            self.command_id = send_command("powershell.exe",[])
+          end
+    
+          def with_command_shell(input, arguments = [])
+            tries ||= 2
+            send_stdin(input)
+            yield shell_id, self.command_id
+          rescue WinRM::WinRMWSManFault => e
+            raise unless FAULTS_FOR_RESET.include?(e.fault_code) && (tries -= 1) > 0
+    
+            reset_on_error(e)
+            retry
+          end
+    
+          def cleanup_shell
+            cleanup_command(command_id)
+          end
+    
+          # Runs the specified command with optional arguments
+          # @param block [&block] The optional callback for any realtime output
+          # @yieldparam [string] standard out response text
+          # @yieldparam [string] standard error response text
+          # @yieldreturn [WinRM::Output] The command output
+          def read_stdout(&block)
+            open unless shell_id
+            create_proc unless command_id
+            begin
+              response_reader.read_output(command_output_message(shell_id, command_id), &block)
+            rescue WinRM::WinRMWSManFault => err
+              # If no output is available before the wsman:OperationTimeout expires,
+              # the server MUST return a WSManFault with the Code attribute equal to
+              # 2150858793. When the client receives this fault, it SHOULD issue
+              # another Receive request.
+              # http://msdn.microsoft.com/en-us/library/cc251676.aspx
+              if err.fault_code == '2150858793'
+                yield nil, nil
+              else
+                raise
+              end
+            end
+          end
+    
+          def send_stdin(input)
```

Comment:
If we're providing our own overrides to `ReceiveResponseReader`, would it make sense to match the expected API in command shell and set rstream to be the `ReceiveResponseReader` object?

i.e. Updating this class to have a method for `write` and `get_once`. This would allow us to update `winrm_command_shell` to no longer override the base class functions, and use the normal implementation as defined here:

https://github.com/rapid7/metasploit-framework/blob/2f0509f2b56dbb73f28a19d5956de22c0bde0e91/lib/msf/base/sessions/command_shell.rb#L663-L673

Reply:
Agree that it'd be nice to have stream-like behaviour to minimise the code duplication; although `ReceiveResponseReader` probably isn't the correct place, as it's solely responsible for reading responses (not sending).

I split the stream-like behaviour of the command shell object into its own object (`WinRMStreamAdapter`) so we can let the base class use `rstream` in the cases you describe.

https://github.com/smashery/metasploit-framework/blob/winrm_improvements/lib/msf/base/sessions/winrm_command_shell.rb#L13-L144